Lut

jammont opened the pull request:
# Description

Refactoring the LUT code to standardize RT outputs and file formats. 

More details to be added.

pgbrodrick committed:
started work, lots to go!

pgbrodrick committed:
made it through 6s, libradtran, modtran.  LUT is next for RTE integration

unbohn committed:
Added contents from look_up_tables.py to radiative_transfer_engine.py.

jammont committed:
WIP generalized HDF5 handler functions, plus some modtran utility functions inspired by Nimrod's research

jammont committed:
Added mypy

jammont committed:
Added xarray versions as alternatives

unbohn committed:
Merge remote-tracking branch 'james_fork/universal_lut' into lut

unbohn committed:
Added LUT reading from HDF5 file to radiative_transfer_engine.py.

pgbrodrick committed:
Merge pull request #384 from pgbrodrick/lut_rte

LUT refactor

pgbrodrick committed:
updates for distribution - needs some thoughts about ray integration w/ classes

NimrodCarmon commented on isofit/utils/luts/channel_read.py:
```diff
@@ -0,0 +1,272 @@
+
+    We parse them one wavelength at a time."""
+
+    multipart_transmittance = multipart
```
We should use re instead here to find the SZA value in the filename. It would be much safer.

We can use the following:

`match = re.search(r'SZA-([\d]+\.\d+)', infile)

    if match:
        SZA = float(match.group(1))
        coszen = np.cos(np.deg2rad(SZA))`

jammont replied:
This has been updated to extract the solar zenith from the tp6 file and take the cosine from there:
```python
file = os.path.join(self.sim_path, self.point_to_filename(point))

solzen = self.load_tp6(f"{file}.tp6")
coszen = np.cos(solzen * np.pi / 180.0)
```

NimrodCarmon reviewed (changes requested):
Just a small change to make the SZA value read from the channel file name safer

jammont committed:
Working example of distributed runSims as part of the RTE class and using Ray

pgbrodrick committed:
rte output name tweaks from team, geometry updates from team

unbohn committed:
Merge pull request #387 from pgbrodrick/lut_names

rte output name updates

pgbrodrick committed:
nomenclature change for clarity

pgbrodrick committed:
remove bound statevector element reference from rte, those belong in rt

pgbrodrick committed:
modified streaming function calls

unbohn committed:
Restructured radiative_transfer.py by adding physics-based functions from the RT engine modules.

unbohn committed:
Merge branch 'isofit:lut' into lut

unbohn committed:
Moved three albedo calculation from modtran.py to radiative_transfer.py.

unbohn committed:
Removed physics-based radiative transfer functions from engine modules.

unbohn committed:
Updated setup of MODTRAN config file to account for three albedo meth…
…od when multipart transmittance is enabled.

unbohn committed:
Changed nomenclature of transmittance quantities for clarity.

jammont committed:
Zarr implementation of lut functions

pgbrodrick committed:
Merge pull request #386 from jammont/luts.nc

NetCDF for LUTs

unbohn committed:
Moved 2-albedo method to radiative_transfer_engine.py.

unbohn committed:
Merge branch 'isofit:lut' into lut

unbohn committed:
Attached two-albedo method as a function to RTE self.

unbohn committed:
Updated radiative_transfer_engine.py for being used for testing.

jammont committed:
Implements NetCDF framework; moved shared utilities to lututils

pgbrodrick committed:
updates to get 6s and mdotran adapted, probably not working yet

pgbrodrick committed:
Merge pull request #391 from jammont/luts.nc

LUT NetCDF part 2

pgbrodrick committed:
Merge pull request #390 from unbohn/lut

Refactoring of physics-based RT functions.

unbohn committed:
Resolved merge conflicts from upstream LUT branch.

unbohn committed:
Merge branch 'rte_ind' into lut_tests

# Conflicts:
#	isofit/radiative_transfer/radiative_transfer_engine.py

unbohn committed:
Added a simple first test case for building and reading a LUT file.

pgbrodrick committed:
Merge pull request #392 from unbohn/lut_tests

RT engine and LUT tests.

pgbrodrick committed:
slight restructure to prevent circular import

pgbrodrick committed:
path and wavelength updates in rte

jammont committed:
Added point dimension values

jammont committed:
Merged

jammont committed:
Merged

jammont committed:
Merge branch 'lut' of github.com:isofit/isofit into lut

jammont committed:
Reworked creation of RTEs in RT to pass parameters easier

jammont committed:
RTEs are now successfully initializing zarr stores

jammont committed:
ForwardModel successfully initializes

jammont committed:
Isofit.run() creates a FM and begins inversions until a geometry exception now

jammont committed:
Fixed isofit root path resolving

jammont committed:
Fixed missing config param treat_as_emissive

jammont committed:
Fixes connections between sims and luts

jammont committed:
Simulated modtran simulations to work around inability to run real simulations

jammont committed:
Works with both prebuilt and empty luts now

jammont committed:
Building interpolators now seems to work

jammont committed:
Interpolators build and usable via temporary new rte.get()

jammont committed:
Added topography_model to RT; re-added summarize() to RTE; temporarily added extra keys to rte.get return

jammont committed:
Isofit(config).run() working first pass

jammont committed:
Fixes to running without a lut store prebuilt

jammont commented:
This branch is now capable of the following:
- Initialize the `Isofit` object, which initializes `ForwardModel` and that a `RadiativeTransfer` and that one or more `RadiativeTransferEngines`
  - `ModtranRT` is the only RTE that has been tweaked enough to work with the new architecture
- `Isofit(config).run()` working with two Pasadena example files (darklot and multimodtran)

This is still very much a work in progress.

---

Issues that still need to be addressed:
- [ ] `radiative_transfer.py#L109` - `topography_model` is retrieved from each engine config, should RT bother with this param? It's used in a few RT functions, though
- [ ] `radiative_transfer_engines.py#L275` - `get()` I dropped the `geom` object as I couldn't quite understand the goals of the previous `get` function. I understand it was inspired by attempts to simplify sRTMnet's `.get()`, but it didn't seem to work with the modtran examples I was using. The alt version I wrote here seems to work 🤷 
  - [ ] Same function, as we've changed the key names of the LUT file several of the downstream functions break. Known keys are [`transm`, `transup`]. I'm not sure if this is how you guys would like to handle this or if the downstream functions (`isofit.inversion.inverse_simple.py`) should be updated to reflect the new LUT structure. Handling this way temporarily alleviates the issue at the cost of negligible overhead, but it may not be correct.
- [ ]` radiative_transfer_engines.py#L211` - Should we build an interpolator for every key in the LUT? Or is there a config option to limit which to build/load?
- [ ] `radiative_transfer_engines.py#L209` - Need support for reading/writing interpolator pickles
- [ ] `radiative_transfer_engines.py#L85` - Where does the `solar_irr` and `coszen` come from? Seems like the individual RTEs, but I temporarily stepped around the modtran simulation issues so I might just not be hitting the lines to set these. For now I just hardcoded them to `[1]`
- [ ] `modtran.py#L725` - Temporarily created a dummy class to simulate these two functions that `ModtranRT` needs implemented. `read_simulation_results` is likely just `ModtranRT.load_chn`, but `make_simulation_call` may need some real work
- [ ] All configs need to tweak their engine sections with:
  - [ ] Need to develop a `VerifyZarr(path)` function. For documentation's sake, `lut_path` needs to be a path to a zarr store which is a directory. Since in previous ISOFIT versions these were directories for us too, the pipeline will raise a vague `FileNotFoundError` if previous version LUT stores exist when really it's just invalid. 
  - [ ] New key `interpolator_base_path`, currently unused
- [ ] Need to refactor other RTE subclasses to support the new architecture:
  - [ ] sRTMnet
  - [ ] 6S
  - [ ] LibRadtran

I'll update this comment as needed.

pgbrodrick committed:
working through basis to address James comments on github, incomplete...james to clean up

jammont committed:
Fixed RTE init order of operations and reader funcs of modtranrt

jammont committed:
Added engine_config.sim_dir which is the old lut_dir which was repurposed to lut_path now that luts and sims diverge

jammont committed:
Reworked netcdf initialization as well as implemented updatePoint to actually work by point values only instead of point index

jammont committed:
ModtranRT v2.0: Fully working reimplementation that simplifies the reader functions

jammont committed:
Fixed netcdf.updatePoint to not assume dim ordering, updated RTE to start using netcdf instead of zarr

jammont committed:
NetCDF now works

jammont committed:
Moved old two_albedo_method to ModtranRT instead of RTE

jammont committed:
ModtranRT now initializes

jammont committed:
Added luts.extractPoints, load by existing lut file appears to be working again

jammont committed:
Edge case handling retrieving bad key

jammont committed:
Fixed bad reference to rte.lut_grids

jammont committed:
Fixed interpolators

jammont committed:
Added support for storing constants, single dimension, and multi dimensional variables in the lut file; added coszen, solzen to lut

jammont committed:
Isofit(config).run() now completes through inversions - Testing needed

jammont committed:
Fixed ModtranRTv1 to now work; Fixed ModtranRTv2 to now produce indentical outputs to v1

jammont committed:
Transpose luts to (N points, M wavelengths)

jammont committed:
Moved two_albedo_method back to rt.py

unbohn committed:
Tackled the modtran.py non-multipart (classic) transmission read by adjusting the read so that transm_down_dif gets populated (transm by itself no longer being a key).

unbohn committed:
Merge remote-tracking branch 'origin/lut' into lut

# Conflicts:
#	isofit/radiative_transfer/modtran.py

unbohn committed:
Moved two_albedo_method back to radiative_transfer_engine.py.

unbohn committed:
Updated get function of radiative_transfer_engine.py.

jammont committed:
Reworked channel parsing to support 1, 2, or 3 multipart transmittances

jammont committed:
Added function to extract nd shape of lut for VectorInterpolator optimization

jammont committed:
Merge branch 'lut' of github.com:isofit/isofit into lut

jammont committed:
Fixed reshaping lut from 2d to nd for VectorInterpolator; fixed modtran.two_albedo_method v1 and v2 clashing; removed unused netcdf function

jammont committed:
Changed key names in inverse_simple to match Niklas' changes; Refactored get and _lookup_lut functions, should now be working and stable

jammont committed:
Isofit(config).run() back to a working state - Testing needed

unbohn committed:
Further adjusted LUT transmittance key names and added more clarity to two_albedo_method, including docstrings and (hopefully) catchy variable names.

unbohn committed:
Merge remote-tracking branch 'origin/lut' into lut

# Conflicts:
#	isofit/inversion/inverse_simple.py
#	isofit/radiative_transfer/modtran.py
#	isofit/radiative_transfer/radiative_transfer_engine.py

jammont committed:
Updated Pasadena example configs for v3

jammont committed:
Default some lut keys to zero instead of nan to support single transmittance cases

jammont committed:
Removed dependency on dask

unbohn committed:
Updated environment_isofit_basic.yml and setup.cfg.

unbohn commented:
@pgbrodrick @jammont I updated environment and setup files since a few packages were missing, which are needed within our new lut branch. They include `xarray`, `netCDF4`, and `dask`.

Although being removed from `luts/netcdf.py` in [11b5636](https://github.com/isofit/isofit/pull/383/commits/11b56362703c339eac34fb3978ce2e586bd722dd), `dask` is still needed by `xarray` as shown in the following traceback:

../luts/netcdf.py:71: in initialize
    ds.to_netcdf(file, mode="w", compute=False)
/Users/bohn/mambaforge/envs/isofit_lut_env/lib/python3.10/site-packages/xarray/core/dataset.py:2279: in to_netcdf
```
    def to_netcdf(
    ...
    if not compute:
       import dask
    ...
```
ModuleNotFoundError: No module named 'dask'

However, adding `dask` as a conda package to the environment file works perfectly fine.

A more severe problem was actually caused by `tensorflow`. I created a fresh ISOFIT environment based on the `environment_isofit_basic.yml`, where `tensorflow` is part of the `pip` dependencies. Then I got the following error when trying to import the module:

`Illegal instruction: 4`

Haven't seen this before, but it's somehow related to the CPU infrastructure of my Mac. `tensorflow` requires CPUs that have AVX instruction sets, which should be the case for my M1 since it worked before. No clue what has changed. Anyhow, moving `tensorflow` to the conda dependencies of the environment file solved the issue. I could imagine that this is somewhat related to previous issues we had (see #365 and #366). I'm not sure yet if we can leave it this way or run into other installation issues with `tensorflow` again. Let me know what you think!

unbohn committed:
Adjusted name of conda environment in environment_isofit_basic.yml.

unbohn committed:
Fixed typo in ndsplines package version.

pgbrodrick commented:
@unbohn / @jammont - can you test if the non-gpu version of pytorch works cleanly?  We are rebuilding the emulator and have the opportunity to switch, and my understanding is that the CPU implementation of pytorch is a fairly clean install.  If you can confirm that it resolves the above issues (across platforms), we can make the transition, as tensorflow seems to be causing consistent challenges.

unbohn commented:
@pgbrodrick Yes, I'll take a look! Would be a relief if we could get rid of `tensorflow`.

unbohn commented:
@pgbrodrick @jammont I tested the `pytorch` installation on both `osx-64` and `linux-64` architectures, and can confirm that it works pretty cleanly. Adding `pytorch` to the conda-forge based `environment_isofit_basic.yml` leads to an installation of the non-gpu version by default. The gpu version is only available on the pytorch conda channel. So, we're safe as long as we stick to conda-forge only. I also tested importing and running a few simple tests with `torch` on both architectures - no issues at all!

pgbrodrick committed:
bring simulation call down

jammont committed:
SixSRT Refactor

jammont committed:
Merge branch 'lut' of github.com:isofit/isofit into lut

jammont committed:
Incremented version to 3.0 for testing

jammont committed:
Fixed script writelines

jammont committed:
SixSRT now builds with SantaMonica example

jammont committed:
Added a preSim function in prep for sRTMnet refactor

jammont committed:
WIP sRTMnet v2

jammont committed:
Added postSim func; tweaked SixSRT to use postSim

jammont committed:
Fixed elusive pickling bug with RTE by disabling __getattr__ altogether; SixSRT now back to building as before

jammont committed:
Fixes to SixSRT; added better support for wl,fwhm overrides in RTE

jammont committed:
sRTMnet refactor draft

jammont committed:
sRTMnet now builds

jammont committed:
Moved ModtranRT.__init__ to preSim instead, merged v1 with v2; verified building again

jammont committed:
Removed libradtran support

jammont committed:
Fixed pytests; removed more libradtran references

jammont committed:
SixSRT exits if both input and output files exists; fixed SixSRT not having esd available in existing-lut cases

jammont committed:
Cleanup

jammont committed:
Removed libradtran and fixed example for 3.0

jammont committed:
Updated SantaMonica configs for v3

jammont committed:
Fixed SixSRT support for loading from existing lut; added more properties to RTE

jammont committed:
Added force-reinstall to worfklows to hopefully bugfix tests failing because old version still present

jammont committed:
Applied Black

jammont committed:
Updated workflows to temporarily manually install new dependencies

jammont committed:
Switched workflow from bash to python version

jammont committed:
Few bugfixes to apply_oe, still broken

jammont committed:
Fixed SixSRT mishandling NaNs; improved logging to indicate when a parsing error occurs

jammont committed:
Updated deprecated variables

jammont committed:
Updated SixSRT log.error statement

jammont committed:
Improved logging

jammont committed:
Fixed checking if override values are provided

jammont committed:
Improved logging

jammont committed:
Improved logging

jammont committed:
Improved logging

jammont committed:
Improved logging

jammont committed:
Improved logging

jammont committed:
Improved logging

jammont committed:
Improved logging

jammont committed:
Bugfixes for EMIT

jammont committed:
Improved updatePoint failure catching and race conditions

jammont committed:
Backwards support for old geom key names plus updated apply_oe with new names

jammont committed:
Updated geom key index value

jammont committed:
Fixed RTE.get accessing geom keys using geom indices

jammont committed:
Remove statevector_names from geometry_input_names before determining indices

jammont committed:
Fixed RTE.indices to work with geom keys properly now; Updated apply_oe with new key names

jammont committed:
Moved log statement

jammont committed:
Replaced netcdf updatePoint restarts with temporary mutex to stagger io

jammont committed:
sRTMnet should now skip makeSim; swapped stacking order for emulator input (experimental)

jammont committed:
Updated SixSRT geom keys to v3 names

jammont committed:
Undo 6S key name changes for now

jammont committed:
Split lines

jammont committed:
Fixed wrong irradiance_current variable

pgbrodrick committed:
update GNDALT hardcoded reference

jammont committed:
Removed look_up_tables.py and fixed any dependencies

jammont committed:
Removed research files, moved luts/netcdf.py to utils/luts.py, updated imports to match

jammont committed:
Merge branch 'lut' of github.com:isofit/isofit into lut

jammont committed:
Moved luts.py from utils to radiative_transfer to avoid the circular import issue #334

jammont committed:
Created staticmethod parse_file to give greater modularity and debugging control of SixSRT

pgbrodrick committed:
update 6s input keys

pgbrodrick committed:
hanging surface_elev_km 6s key update

jammont committed:
Fixed example in docstring

jammont committed:
Merge branch 'lut' of github.com:isofit/isofit into lut

jammont committed:
Fixed import locations

jammont committed:
Added logging statements

pgbrodrick committed:
make elevation negatives in 6s safer

jammont committed:
Added translation for ModtranRT names, removed rebuild_cmd (replaced by makeSim); updated SixSRT comments to reflect changes

jammont committed:
Added optional param to makeSim to restore similar functionality of the old rebuild_cmd

jammont committed:
Fallback to self.lut_names when engine_config.statevector_names is not provided

unbohn committed:
Fixed typo in surface elevation key.

jammont committed:
Added NaN handler for sRTMnet

jammont committed:
Merge branch 'lut' of github.com:isofit/isofit into lut

jammont committed:
Moved quantity selection around

jammont committed:
Fixed wrong variable reference

unbohn committed:
Switched MODTRAN configuration from IPARM 11 to 12.

unbohn committed:
Readded GMTIME and TRUEAZ to MODTRAN template since these keys are used by sRTMnet.py.

unbohn committed:
Adjusted minimum spacing for relative azimuth angle.

jammont committed:
Added new dependencies to the hypertrace workflow

jammont committed:
Implemented LUT subsetting

unbohn committed:
Updated RTM subprocess calls to capture and display stdout error messages.

unbohn committed:
Merge remote-tracking branch 'origin/lut' into lut

jammont committed:
Added netcdf reader to fileio.SpectrumFile

jammont committed:
Merge branch 'lut' of github.com:isofit/isofit into lut

unbohn committed:
Updated arguments for building MODTRAN template in hypertrace workflow.

unbohn committed:
Added regeneration of MODTRAN input wavelength file to preSim function in modtran.py.

jammont committed:
Added write support for netcdf in fileio

jammont committed:
Merge branch 'dev' of github.com:isofit/isofit into lut

jammont committed:
Fixed merge conflicts

unbohn committed:
Adjusted path to MODTRAN wavelength filter file to use sim_path instead of lut_path.

jammont committed:
Added back missing variables

pgbrodrick committed:
Merge pull request #434 from jammont/v2to3

Merge v2 into LUT

jammont committed:
Added subsetting examples to luts.load

unbohn committed:
Removed os.path.dirname to fix directory to MODTRAN wavelength filter file.

unbohn committed:
Merge remote-tracking branch 'origin/lut' into lut

jammont committed:
Reworked wavelength handling from file

jammont committed:
Merge branch 'lut' of github.com:isofit/isofit into lut

unbohn committed:
Improved handling of wavelength information.

jammont committed:
Reworked subsetting to now use lut_names instead of a new config param as well as broke it out into its own function

jammont committed:
merge

jammont committed:
Merge branch 'lut' of github.com:isofit/isofit into lut

unbohn committed:
Fixed Santa Monica example workflow.

unbohn committed:
Fixed linting errors.

unbohn committed:
Replaced scipy with numpy in glint surface model.

unbohn committed:
Pinned xarray to avoid TypeError.

unbohn committed:
Renamed ISOFIT environment.

unbohn committed:
Pinned installation of xarray in example-testing.yml.

unbohn committed:
Pinned installation of xarray in all workflows.

unbohn committed:
Added lut_subset key to config files of Pasadena example.

unbohn committed:
Removed some docstrings from workflow .yml files.

unbohn committed:
Added pinning of xarray to comprehensive pytests.

unbohn committed:
Added some adjustments to the hypertrace workflow.

jammont committed:
Changed print to logger

jammont committed:
Merge branch 'lut' of github.com:isofit/isofit into lut

unbohn committed:
Renamed lut_subset to lut_names again and fixed workflows accordingly.

jammont committed:
Fixed lut_subset config to be lut_names; fixed dim ordering for the point dim in load

jammont committed:
Merge branch 'lut' of github.com:isofit/isofit into lut

unbohn committed:
Final fixes to the hypertrace workflow.

unbohn committed:
Merge remote-tracking branch 'origin/lut' into lut

unbohn committed:
Fixes to comprehensive pytest workflows.

unbohn committed:
Removed code artifacts.

jammont committed:
Added subsetting support for interpolating to a point and encompassing with gte/lte keys

unbohn committed:
Enabled multipart transmittance mode for 6s RTM.

unbohn commented:
@pgbrodrick Commit [3672faa](https://github.com/isofit/isofit/pull/383/commits/3672faa7653873f9c61bd0b5b18ab120e627e3ea) implements the multipart transmittance for 6s. Currently, the code doesn't separate diffuse and direct upward transmittances, but neither the topography model nor the glint model need that. I ran several tests with varying atmospheric parameters and it seems to work well. However, please let me know if you run into any issues!

unbohn commented on isofit/radiative_transfer/sRTMnet.py:
```diff
@@ -215,12 +215,8 @@ def build_sixs_config(engine_config):
-        data["SURFACE"]["GNDALT"] * 1000.0,
-    )
+    solar_azimuth = data["GEOMETRY"]["PARM1"]
+    solar_zenith = data["GEOMETRY"]["PARM2"]
```
@pgbrodrick I removed the calculation of the sun position here as both solar azimuth and zenith angles are now directly provided by the updated MODTRAN template (using IPARM 12).

unbohn committed:
Fixed a small bug.

jammont committed:
Added option to use dask or not; added warning for NaNs in subselected lut

jammont committed:
Added utility script to out-of-core process 6S outputs to resample and predict with an sRTMnet emulator

jammont committed:
Merge branch 'lut' of github.com:isofit/isofit into lut

unbohn committed:
Added resampling option to prebuilt LUT file.

pgbrodrick committed:
Integrate external netcdf with apply_oe (#439)

- removes multisurface_oe (to be re-integrated later)
- migrates core apply_oe functions to template_construction
- sets up pass-throughs to allow for an external netcdf to be referenced in apply_oe
- re-organizes apply_oe
- applies some patch fixes to lut.py

pgbrodrick committed:
handle merge conflicts

unbohn committed:
Merge pull request #441 from pgbrodrick/lut_merge

Mege dev into lut

pgbrodrick committed:
handle merge conflicts...again

pgbrodrick committed:
update readme

pgbrodrick commented:
This marks the beginning of ISOFIT 3x!
